packages/mendel-pipeline/src/step/gst.js

```diff
@@ -0,0 +1,174 @@
+                    transformIds: ['raw'],
+                    source,
+                    deps,
+                });
+            },
+            removeEntry: (entry) => {
+                this._registry.removeEntry(entry.id);
+            },
+        };
+    }
+
+    gstDone(entry) {
+        this._processed.add(entry.id);
+        if (this._processed.size >= this._cache.size()) {
+            this._processed.clear();
+            if (++this._curGstIndex >= this._gsts.length) {
+                this._cache.entries().forEach(({id}) => {
+                    this.emit('done', {entryId: id});
+                });
+            } else {
+                this._cache.entries().forEach(entry => this.performHelper(entry));
+            }
+        }
+    }
+
+    // this is conforming to the steps API
+    perform(entry) {
+        if (this._gsts.length <= this._curGstIndex) return this.gstDone();
+        this.performHelper(entry);
+    }
+
+    performHelper(entry) {
+        const prevExec = this._gsts.slice(0, this._curGstIndex);
+        const curExec = this._gsts.slice(0, this._curGstIndex + 1);
+        const prevTransformIds = this._registry.getClosestPlanTransformIds(entry.id, prevExec);
+        const nextTransformIds = this._registry.getClosestPlanTransformIds(entry.id, curExec);
+        const proxy = EntryProxy.getFromEntry(entry, prevTransformIds);
+        const currentGstConfig = this._transforms.find(({id}) => id === this._gsts[this._curGstIndex]);
+        const currentGst = require(currentGstConfig.plugin);
+
+        // If no GST is planned for this type, abort.
+        // If plugin doesn't want to deal with it, abort.
+        if (this._processed.has(entry.id) || !nextTransformIds.length || !currentGst.predicate(proxy)) {
+            return this.gstDone(entry);
+        }
+
+        const graph = this._registry.getDependencyGraph(entry.normalizedId, (depEntry) => {
+            const xformIds = this._registry.getClosestPlanTransformIds(depEntry.id, prevExec);
+            const dependecyMap = depEntry.getDependency(xformIds);
+            return Object.keys(dependecyMap).map(literal => {
+                // TODO remove the dep property
+                return dependecyMap[literal].main;
+            });
+        });
+
+        this.explorePermutation(graph, (chain, variation) => {
+            const [main] = chain;
+            this._processed.add(main.id);
+
+            // We need to create proxy for limiting API surface for plugin writers.
+            const context = this.getContext(nextTransformIds);
+            const chainProxy = chain.map(dep => {
+                const transformIds = this._registry.getClosestPlanTransformIds(dep.id, prevExec);
+                return EntryProxy.getFromEntry(dep, transformIds);
+            });
+            const result = currentGst.transform(context, chainProxy);
+            const [proxiedMain] = chainProxy;
+            proxiedMain.id = convertVariationalPath(
```

Comment:
This creates rsync like main file.

If base has a main that variation does not have, this creates the virtual entry. I might just add virtual entry instead later.